src/masternodes/masternodes.cpp

```diff
@@ -917,73 +917,129 @@ CAmount CCollateralLoans::precisionRatio() const
+    if (!res.ok)
+        return std::move(res);
+
+    LogPrint(BCLog::LOAN, "\t\t%s(): totalCollaterals - %lld, totalLoans - %lld, ratio - %d\n",  
+        __func__, result.totalCollaterals, result.totalLoans, result.ratio());
+    return ResVal<CCollateralLoans>(result, Res::Ok());
+}
+
+ResVal<CAmount> GetValidatedIntervalPrice(CCustomCSView *view, CTokenCurrencyPair priceFeedId, bool useNextPrice, bool requireLivePrice) {
+    auto tokenSymbol = priceFeedId.first;
+    auto currency = priceFeedId.second;
+    
+    LogPrint(BCLog::ORACLE,"\t\t%s()->for_loans->%s->", __func__, tokenSymbol); /* Continued */
+
+    auto priceFeed = view->GetFixedIntervalPrice(priceFeedId);
+    if (!priceFeed)
+        return std::move(priceFeed);
+    
+    if (requireLivePrice && !priceFeed.val->isLive(view->GetPriceDeviation())) {
+        return Res::Err("No live fixed prices for %s/%s", tokenSymbol, currency);
     }
 
-    for (const auto& col : collaterals.balances) {
-        auto collToken = HasLoanSetCollateralToken({col.first, height});
-        if (!collToken)
-            return Res::Err("Collateral token with id (%s) does not exist!", col.first.ToString());
-        LogPrint(BCLog::ORACLE,"\t\t%s()->for_collaterals->id(%s)->", __func__, collToken->idToken.ToString()); /* Continued */
-        auto priceFeed = GetFixedIntervalPrice(collToken->fixedIntervalPriceId);
-        if (!priceFeed)
-            return std::move(priceFeed);
-        if (requireLivePrice && !priceFeed.val->isLive(GetPriceDeviation()))
-            return Res::Err("No live fixed prices for %s/%s", collToken->fixedIntervalPriceId.first, collToken->fixedIntervalPriceId.second);
-        auto price = priceFeed.val->priceRecord[int(useNextPrice)];
-        if (price <= 0)
-            return Res::Err("Negative price (%s/%s)", collToken->fixedIntervalPriceId.first, collToken->fixedIntervalPriceId.second);
-        auto amount = MultiplyAmounts(price, col.second);
-        if (price > COIN && amount < col.second)
-            return Res::Err("Value/price too high (%s/%s)", GetDecimaleString(col.second), GetDecimaleString(price));
-        amount = MultiplyAmounts(collToken->factor, amount);
-        auto prevCollaterals = ret.totalCollaterals;
-        ret.totalCollaterals += amount;
-        if (prevCollaterals > ret.totalCollaterals)
-            return Res::Err("Exceed maximum collateral");
-        ret.collaterals.push_back({col.first, amount});
+    auto priceRecordIndex = useNextPrice ? 1 : 0;
+    auto price = priceFeed.val->priceRecord[priceRecordIndex];
+    if (price <= 0)
+        return Res::Err("Negative price (%s/%s)", tokenSymbol, currency);
+
+    return ResVal<CAmount>(price, Res::Ok());
+}
+
+Res PopulateLoansData(CCustomCSView *view, CCollateralLoans& result, CVaultId const& vaultId, 
+    uint32_t height, int64_t blockTime, bool useNextPrice, bool requireLivePrice) {
+    if (!view)
+        return Res::Err("Unexpected call without chain state");
```

Comment:
Remove that check.

Reply:
Unless we change the above or turn it into private, this is not ideal. But can turn into private. Will do that.